docs/docs/proposals/003-grafanaserviceaccount-crd.md

```diff
@@ -0,0 +1,84 @@
+---
+title: "GrafanaServiceAccount CRD"
+linkTitle: "GrafanaServiceAccount CRD"
+---
+
+## Summary
+
+Add a GrafanaServiceAccount CRD so the operator can create Grafana Service Accounts using CRs to automate the setup of Grafana Service Accounts in newly setup grafana instances. 
+
+Today its required to manually set them up in a running grafana instance using the Grafana GUI or the HTTP-API. This document introduces the suggestion of having them as separate objects that can be setup by the operator on deploy.
+
+The suggested new features are:
+* Let the operator create a grafana service account using a GrafanaServiceAccount CRD.
+* Let the operator store the token as a k8s-secret
+
+## Info
+
+Status: Suggested
+
+## Motivation
+
+Today Grafana Service Accounts has to be created after deploy when the grafana is running using the HTTP-API or the GUI. I instead suggest to have a Grafana Service Account CRD so that the Service Accounts could be predefined and created by the operator at deploy and the tokens will be created as k8s-secrets that then can be read by applications when needed.
+
+## Verification
+
+- Create integration tests for the operator creating grafana service accounts from a bare minimum yaml
+- Create integration tests for the operator creating grafana service accounts from a fully specified yaml
+- Create integration tests to check that it can rotate/invalidate tokens with TTL set (and passed).
+
+## Current solution
+
+Currently you are only able to create these grafana service accounts using the grafana GUI or by using the HTTP-API after the grafana has already been deployed and is running. And its removed when the grafana pod is restarted/redeployed without persistent storage. Meaning a new service account has to be manually created and its new token has to be updated where its being used.
+
+## Proposal
+
+My proposal is to handle grafana service account as a separate CRD that can be specified by the user even before setup and that can be included in a CICD pipeline. It will enable so that the operator can create predefined service accounts on deploy and store the token in a k8s-secret readable by other applications without any manual steps.
+
+### Defining what Grafana Service Account belongs to what grafana.
+I suggest that its done in the same way as dashboards and datasources are using instanceSelector to find the proper grafana, the same thing could be done with the grafana service accounts using the same labels etc.
```

Comment:
+1 to instanceSelector, however, how would we handle the generation of the GSA token secret? Seeing as we will create a one-to-many relationship here, that means one-secret per selected `Grafana`

A number of other questions sprawl up for me:
1. What namespace do we create the resulting `grafana-sa-token-secret-<identifier>` in? woould it be the the namespace of the `Grafana` we create it for, or the namespace within which the `GrafanaServiceAccount` lives?

2. How will we reference all the secrets from within the `GrafanaServiceAcccount` CR? looking at this from the use case of: if a user deletes the `GSA` how will the operator be able to delete all subsequent serviceaccounts on all managed `Grafana` instances, as well as their relevant secrets.

Reply:
Id say that if multiple tokens are specified in the grafanaserviceaccount cr then the operator would create those tokens as k8s-secrets with the name that is specified in the grafanaserviceaccount cr and that will be the reference for grafana what secret is used for what grafanaserviceaccount. The default would only be to create one with the default name, if spec.create.generateTokenSecret is true that is. Which should be default.

As for the other questions,

1. I think it's preferable to have the secret residing in the same namespace as the Grafana CR due to how RBAC rules for secrets are often more restricted in the view scope, and a normal usage pattern would be to use said secret for other applications.
2. I think that with the k8s-secrets containing the tokens as a list under spec.tokens the operator could remove the secrets when the GSA is removed. I could also see the use of a "Managed by: Grafana-operator"-label on the secrets, allowing the operator to find all its created secrets and compare if they belong to any GSA that still exists.

Reply:
To address the security concerns outlined by @NissesSenap  & deletion mapping, I have the following idea:

Instead of handling serviceaccounts as separate resources, we could add them as a field to grafana instances directly. This would solve the security concers by restricting service account creation to the same group of users which can already modify the Grafana instance while also keeping everything in one place.

What are your takes on this?

Reply:
The good thing is that it's simple. The bad thing is that if you want a service account, the owner of the grafana instance will have to create it.

My guess is that this will be the case most of the time any way, so it's probably not a huge issue.
What do you think @MickeHedlund ?

Reply:
I dont see an issue with the owner of the grafana having to create service accounts. That how we would want it to work in our use cases.

Reply:
It would solve lots of issues and make us having to write one less controller which is always nice.
@MickeHedlund could you update the proposal to match this change?
More or less the same, but move everything over to the `Grafana` CR

Reply:
I will update it, I will have to see if I have time to do it today, if not it will be updated early next week.

Reply:
I have pushed an updated document, let me know if I need to correct something.